README.md
# runit

This is a fork of the runit init/services management framework originally 
developed by Gerrit Pape.  

To quote the original website: 
> runit is a cross-platform Unix init scheme with service supervision, a replacement 
> for sysvinit, and other init schemes. It runs on GNU/Linux, *BSD, MacOSX, Solaris, 
> and can easily be adapted to other Unix operating systems."

It is the primary init solution for many Linux distributions, and the optional (and 
often preferred, even though optional) one for many more.

The purpose of this fork is several-fold:

1. runit, while it works well enough for those distributions using it, hasn't had a real update in a while.
2. While it builds well enough, it is not geared to readily package into many distributions- it is packaged as a slashpackage.
3. While it builds well enough, it's build system is not geared for cross-compilation.
4. There needs to be a relatively credible alternative to systemd, this is one of the same.
5. While it is complete, and very capable, there are some aspects of fault tolerance not fully covered as well as they could be.

It is the intention of the long time Linux "Plumbers" that have undertaken this work to 
accomplish the above, providing a solid ecosystem around runit to provide what is believed 
to be a better answer to init/services management, that while it is differing from the 
sysv init structure that has gone before, is in keeping with the UNIX/*BSD/Linux philosophy
of doing only init/services management, leaving _*other*_ programs/schemes to handle udev, 
syslog, etc.

All decisions undertaken by this project are to be done by general consensus of the participants 
of this project.  Designs aren't to all be designed by committee, but any design decisions 
desired to be implemented into mainline is to be agreed upon by the members of this project.



CHANGES
2.1.3
Tue, 15 Aug 2017
  * Reworked directory structure for more maintainability.
  * Reworked build system to use CMake instead of the one previously used.

2.1.2
Sun, 10 Aug 2014 18:01:54 +0000
  * doc/index.html: reorder, cleanup links to distributions; add Linux
    from Scratch, Finnix, VServer.
  * sv.c: properly format status command's output on failure cases.
  * sv.c: support optional LSB init script actions reload and
    try-restart.
  * man/sv.8: "sv exit" does not send a TERM signal to the log service
    (thx Jonathan Nieder).
  * fmt_ptime.c: 64 bits time_t fix for svlogd (tnx Jérémie
    Courrèges-Anglas).
  * sv.c: fix typo that may lead to wrong output from sv when reporting
    status of multiple service directories.
  * doc/index.html: add deepOfix Mail Server to list of distributions
    that include runit (thx Debayan Banerjee).
  * doc/useinit.html: update description on how to use with upstart
    (thx Gabriel de Perthuis).
  * doc/index.html: add Dragora GNU/Linux to list of distributions that
    use runit as default init scheme (thx Matias A. Fonzo).
  * doc/index.html: add ArchLinux to list of distributions that include
    runit (thx Kevin Berry).

2.1.1
Sun, 04 Oct 2009 20:28:38 +0000
  * doc/upgrade.html: fix typo.
  * sv.c: on 'down', send runsv the 'down' command properly if not yet
    done (e.g. when taken up with 'once').

2.1.0
Thu, 24 Sep 2009 22:49:33 +0000
  * pathexec_env.c, pathexec.h: add function pathexec_env_run().
  * chpst.c, man/chpst.8: new option -b argv0: run prog with different
    0th argument.
  * sv.c: fix exit code of status command in lsb mode (thx Mathieu
    Poumeyrol)
  * chpst.c, svwaitdown.c, svwaitup.c: fix checks for return value of
    open_* functions (thx David Reiss).
  * runsv.c: exit with error if [log/]supervise/control exists, but is
    not a fifo (thx Charlie Brady).
  * man/sv.8: clarify which return values apply to the LSB init script
    interface.

2.0.0
Sun, 15 Jun 2008 15:31:05 +0000

1.9.0
Mon, 05 May 2008 22:00:13 +0000
  * doc/upgrade.html: typo.
  * sv.c: service name is also relative to the current directory if it
    ends with a slash.
  * change default directory for services from /var/service/ to /service/.
  * runsv.c: create temporary new status files for log/supervise/ actually
    in log/supervise/.
  * doc/benefits.html: minor additions.

1.8.0
Fri, 21 Sep 2007 00:33:56 +0000
  * man/svlogd.8: add hint on how to manually remove log files after the
    number of log files svlogd should maintain has been reduced; add hint
    about increasing the buffer size if lots of data is to be processed
    in short time.
  * chpst.c, uidgid.c: improve error reporting if getpwnam() or getgrnam()
    fails; cleanup.
  * svlogd.c: don't statically bind local udp socket, but request new
    source port for each log message sent through udp.
  * runit.c: fix typo in error messages (thx Matthew R. Dempsky); minor
    rewording.
  * doc/useinit.html: add instructions on how to use with upstart (thx
    Lloyd Zusman).
  * chpst.c: add explicit braces to avoid ambiguous 'else'.
  * uw_tmp.h1: fallback to UTMP_FILE if _PATH_UTMP is not defined as seen
    on AIX 5.2 ML1, AIX 5.3 ML5 (thx Daniel Clark).
  * sv.c: fix race on check for down if pid is 0 and state is run or finish.
  * runit.c: speed up collecting zombies.
  * runit.c: force check for zombies after a 14 second timeout without
    signals (CHLD, CONT, INT).
  * doc/index.html: typo.
  * doc/*.html: remove $Id$.

1.7.2
Tue, 21 Nov 2006 15:13:47 +0000
  * doc/index.html: add list of distributions that are known to include
    or package runit; some reordering.
  * doc/replaceinit.html: point to faq.html#create, faq.html#tell (thx
    David Kaufman).
  * doc/runscripts.html: point to faq.html#createlog.
  * runit.c: fix arguments to wait_pid() after fork for reboot_system()
    (thx Jan Kampe).
  * man/runsv.8: suggest printf t instead of echo -n t.
  * runsv.c: really don't act on commands in state finish; minor.

1.7.1
Sat, 04 Nov 2006 19:23:29 +0000
  * chpst.c, uidgid.c, uidgid.h: support numerical uid gid through
    chpst -[u|U] :uid:gid...
  * man/chpst.8: adapt; clarify that users and groups are looked up in
    /etc/passwd and /etc/group only.
  * sv.c: properly wait for a service to be restarted on 'restart';
    support checks through -v for pause, cont, kill.
  * doc/runscripts.html: add contributed run scripts (thx Kevin, marlowe,
    ed neville, xavier dutoit).

1.7.0
Sat, 07 Oct 2006 18:24:17 +0000
  * svlogd.c, fmt_ptime.*, man/svlogd.8: new option -ttt: prefix log
    messages with sortable UTC timestamp YYYY-MM-DDTHH:MM:SS.xxxxx.
  * runsv.c, runsv.8: give two arguments to ./finish: exit code and exit
    status of ./run (mostly copied from pipe-tools' npt-supervise, thx
    Laurent Bercot).
  * runit.c: don't make console the controlling terminal for process 1,
    but for stage 1 only; fork before initiating system halt or reboot,
    the linux kernel calls do_exit(0) in kernel/sys.c (thx Jan Kampe).

1.6.0
Thu, 29 Jun 2006 07:52:35 +0000
  * svlogd.c: cleanup *.t files possibly leftover by processor when
    interrupted by signal, also on startup (thx Andras Korn,
    http://bugs.debian.org/369840); 1st rename .t to .s, then remove .u.
  * man/svlogd.8: adapt.
  * svlogd.c: prepend optional timestamp to log messages sent through UDP,
    as done for messages written to dir and to stderr.
  * uidgid.c: properly fixup user/group argument on getpwnam() error.
  * doc/benefits.html: typo; wording.

1.5.1
Wed, 10 May 2006 20:55:35 +0000
  * runsv.c: delay restart of log service at least 1 sec, just as the main
    service.
  * runsvdir.c: don't double-copy filedescriptor.
  * package/README, package/COPYING: 2006.
  * man/runsvctrl.8, man/runsvstat.8, man/svwaitup.8, man/svwaitdown.8,
    doc/runsvctrl.html, doc/runsvstat.html, doc/svwaitdown.html,
    doc/svwaitup.html: remove; obsolete.

1.5.0
Sun, 16 Apr 2006 12:26:50 +0000

src/runsv.c
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <stdio.h>
#include <signal.h>
#include "strerr.h"
#include "error.h"
#include "taia.h"
#include "sig.h"
#include "env.h"
#include "coe.h"
#include "ndelay.h"
#include "fifo.h"
#include "open.h"
#include "lock.h"
#include "iopause.h"
#include "wait.h"
#include "fd.h"
#include "buffer.h"
#include "fmt.h"
#include "byte.h"

#define USAGE " dir"

char *progname;
int selfpipe[2];

/* state */
#define S_DOWN 0
#define S_RUN 1
#define S_FINISH 2
/* ctrl */
#define C_NOOP 0
#define C_TERM 1
#define C_PAUSE 2
/* want */
#define W_UP 0
#define W_DOWN 1
#define W_EXIT 2

struct svdir {
  int pid;
  int state;
  int ctrl;
  int want;
  struct taia start;
  int wstat;
  int fdlock;
  int fdcontrol;
  int fdcontrolwrite;
  int islog;
};
struct svdir svd[2];

int sigterm =0;
int haslog =0;
int pidchanged =1;
int logpipe[2];
char *dir;

void usage () { strerr_die4x(1, "usage: ", progname, USAGE, "\n"); }

void fatal(char *m) {
  strerr_die5sys(111, "runsv ", dir, ": fatal: ", m, ": ");
}
void fatal2(char *m1, char *m2) {
  strerr_die6sys(111, "runsv ", dir, ": fatal: ", m1, m2, ": ");
}
void fatalx(char *m1, char *m2) {
  strerr_die5x(111, "runsv ", dir, ": fatal: ", m1, m2);
}
void warn(char *m) {
  strerr_warn5("runsv ", dir, ": warning: ", m, ": ", &strerr_sys);
}
void warn2(char *m1, char *m2) {
  strerr_warn6("runsv ", dir, ": warning: ", m1, m2, ": ", &strerr_sys);
}
void warnx(char *m1, char *m2, char *m3) {
  strerr_warn6("runsv ", dir, ": warning: ", m1, m2, m3, 0);
}

void stopservice(struct svdir *);

void s_child() { write(selfpipe[1], "", 1); }
void s_term() {
  sigterm =1;
  write(selfpipe[1], "", 1); /* XXX */
}

void update_status(struct svdir *s) {
  unsigned long l;
  int fd;
  char status[20];
  char bspace[64];
  buffer b;
  char spid[FMT_ULONG];
  char *fstatus ="supervise/status";
  char *fstatusnew ="supervise/status.new";
  char *fstat ="supervise/stat";
  char *fstatnew ="supervise/stat.new";
  char *fpid ="supervise/pid";
  char *fpidnew ="supervise/pid.new";

  if (s->islog) {
    fstatus ="log/supervise/status";
    fstatusnew ="log/supervise/status.new";
    fstat ="log/supervise/stat";
    fstatnew ="log/supervise/stat.new";
    fpid ="log/supervise/pid";
    fpidnew ="log/supervise/pid.new";
  }

  /* pid */
  if (pidchanged) {
    if ((fd =open_trunc(fpidnew)) == -1) {
      warn2("unable to open ", fpidnew);
      return;
    }
    buffer_init(&b, buffer_unixwrite, fd, bspace, sizeof bspace);
    spid[fmt_ulong(spid, (unsigned long)s->pid)] =0;
    if (s->pid) {
      buffer_puts(&b, spid);
      buffer_puts(&b, "\n");
      buffer_flush(&b);
    }
    close(fd);
    if (rename(fpidnew, fpid) == -1) {
      warn2("unable to rename pid.new to ", fpid);
      return;
    }
    pidchanged =0;
  }

  /* stat */
  if ((fd =open_trunc(fstatnew)) == -1) {
    warn2("unable to open ", fstatnew);
    return;
  }
  buffer_init(&b, buffer_unixwrite, fd, bspace, sizeof bspace);
  switch (s->state) {
  case S_DOWN:
    buffer_puts(&b, "down");
    break;
  case S_RUN:
    buffer_puts(&b, "run");
    break;
  case S_FINISH:
    buffer_puts(&b, "finish");
    break;
  }
  if (s->ctrl & C_PAUSE) buffer_puts(&b, ", paused");
  if (s->ctrl & C_TERM) buffer_puts(&b, ", got TERM");
  if (s->state != S_DOWN)
    switch(s->want) {
    case W_DOWN:
      buffer_puts(&b, ", want down");
      break;
    case W_EXIT:
      buffer_puts(&b, ", want exit");
      break;
    }
  buffer_puts(&b, "\n");
  buffer_flush(&b);
  close(fd);
  if (rename(fstatnew, fstat) == -1)
    warn2("unable to rename stat.new to ", fstat);

  /* supervise compatibility */
  taia_pack(status, &s->start);
  l =(unsigned long)s->pid;
  status[12] =l; l >>=8;
  status[13] =l; l >>=8;
  status[14] =l; l >>=8;
  status[15] =l;
  if (s->ctrl & C_PAUSE)
    status[16] =1;
  else
    status[16] =0;
  if (s->want == W_UP)
    status[17] ='u';
  else
    status[17] ='d';
  if (s->ctrl & C_TERM)
    status[18] =1;
  else
    status[18] =0;
  status[19] =s->state;
  if ((fd =open_trunc(fstatusnew)) == -1) {
    warn2("unable to open ", fstatusnew);
    return;
  }
  if ((l =write(fd, status, sizeof status)) == -1) {
    warn2("unable to write ", fstatusnew);
    close(fd);
    unlink(fstatusnew);
    return;
  }
  close(fd);
  if (l < sizeof status) {
    warnx("unable to write ", fstatusnew, ": partial write.");
    return;
  }
  if (rename(fstatusnew, fstatus) == -1)
    warn2("unable to rename status.new to ", fstatus);
}
unsigned int custom(struct svdir *s, char c) {
  int pid;
  int w;
  char a[10];
  struct stat st;
  char *prog[2];

  if (s->islog) return(0);
  byte_copy(a, 10, "control/?");
  a[8] =c;
  if (stat(a, &st) == 0) {
    if (st.st_mode & S_IXUSR) {
      if ((pid =fork()) == -1) {
        warn2("unable to fork for ", a);
        return(0);
      }
      if (! pid) {
        if (haslog && fd_copy(1, logpipe[1]) == -1)
          warn2("unable to setup stdout for ", a);
        prog[0] =a;
        prog[1] =0;
        execve(a, prog, environ);
        fatal("unable to run control/?");
      }
      while (wait_pid(&w, pid) == -1) {
        if (errno == error_intr) continue;
        warn2("unable to wait for child ", a);
        return(0);
      }
      return(! wait_exitcode(w));
    }
  }
  else {
    if (errno == error_noent) return(0);
    warn2("unable to stat ", a);
  }
  return(0);
}
void stopservice(struct svdir *s) {
  if (s->pid && ! custom(s, 't')) {
    kill(s->pid, SIGTERM);
    s->ctrl |=C_TERM;
    update_status(s);
  }
  if (s->want == W_DOWN) {
    kill(s->pid, SIGCONT);
    custom(s, 'd'); return;
  }
  if (s->want == W_EXIT) {
    kill(s->pid, SIGCONT);
    custom(s, 'x');
  }
}

void startservice(struct svdir *s) {
  int p;
  char *run[4];
  char code[FMT_ULONG];
  char stat[FMT_ULONG];

  if (s->state == S_FINISH) {
    run[0] ="./finish";
    code[fmt_ulong(code, wait_exitcode(s->wstat))] =0;
    run[1] =wait_crashed(s->wstat) ? "-1" : code;
    stat[fmt_ulong(stat, s->wstat & 0xff)] =0;
    run[2] =stat;
    run[3] =0;
  }
  else {
    run[0] ="./run";
    custom(s, 'u');
    run[1] =0;
  }

  if (s->pid != 0) stopservice(s); /* should never happen */
  while ((p =fork()) == -1) {
    warn("unable to fork, sleeping");
    sleep(5);
  }
  if (p == 0) {
    /* child */
    if (haslog) {
      if (s->islog) {
        if (fd_copy(0, logpipe[0]) == -1)
          fatal("unable to setup filedescriptor for ./log/run");
        close(logpipe[1]);
        if (chdir("./log") == -1)
          fatal("unable to change directory to ./log");
      }
      else {
        if (fd_copy(1, logpipe[1]) == -1)
          fatal("unable to setup filedescriptor for ./run");
        close(logpipe[0]);
      }
    }
    sig_uncatch(sig_child);
    sig_unblock(sig_child);
    sig_uncatch(sig_term);
    sig_unblock(sig_term);
    execve(*run, run, environ);
    if (s->islog)
      fatal2("unable to start log/", *run);
    else
      fatal2("unable to start ", *run);
  }
  if (s->state != S_FINISH) {
    taia_now(&s->start);
    s->state =S_RUN;
  }
  s->pid =p;
  pidchanged =1;
  s->ctrl =C_NOOP;
  update_status(s);
}
int ctrl(struct svdir *s, char c) {
  switch(c) {
  case 'd': /* down */
    s->want =W_DOWN;
    update_status(s);
    if (s->state == S_RUN) stopservice(s);
    break;
  case 'u': /* up */
    s->want =W_UP;
    update_status(s);
    if (s->state == S_DOWN) startservice(s);
    break;
  case 'x': /* exit */
    if (s->islog) break;
    s->want =W_EXIT;
    update_status(s);
    if (s->state == S_RUN) stopservice(s);
    break;
  case 't': /* sig term */
    if (s->state == S_RUN) stopservice(s);
    break;
  case 'k': /* sig kill */
    if ((s->state == S_RUN) && ! custom(s, c)) kill(s->pid, SIGKILL);
    s->state =S_DOWN;
    break;
  case 'p': /* sig pause */
    if ((s->state == S_RUN) && ! custom(s, c)) kill(s->pid, SIGSTOP);
    s->ctrl |=C_PAUSE;
    update_status(s);
    break;
  case 'c': /* sig cont */
    if ((s->state == S_RUN) && ! custom(s, c)) kill(s->pid, SIGCONT);
    if (s->ctrl & C_PAUSE) s->ctrl &=~C_PAUSE;
    update_status(s);
    break;
  case 'o': /* once */
    s->want =W_DOWN;
    update_status(s);
    if (s->state == S_DOWN) startservice(s);
    break;
  case 'a': /* sig alarm */
    if ((s->state == S_RUN) && ! custom(s, c)) kill(s->pid, SIGALRM);
    break;
  case 'h': /* sig hup */
    if ((s->state == S_RUN) && ! custom(s, c)) kill(s->pid, SIGHUP);
    break;
  case 'i': /* sig int */
    if ((s->state == S_RUN) && ! custom(s, c)) kill(s->pid, SIGINT);
    break;
  case 'q': /* sig quit */
    if ((s->state == S_RUN) && ! custom(s, c)) kill(s->pid, SIGQUIT);
    break;
  case '1': /* sig usr1 */
    if ((s->state == S_RUN) && ! custom(s, c)) kill(s->pid, SIGUSR1);
    break;
  case '2': /* sig usr2 */
    if ((s->state == S_RUN) && ! custom(s, c)) kill(s->pid, SIGUSR2);
    break;
  }
  return(1);
}

int main(int argc, char **argv) {
  struct stat s;
  int fd;
  int r;
  char buf[256];

  progname =argv[0];
  if (! argv[1] || argv[2]) usage();
  dir =argv[1];

  if (pipe(selfpipe) == -1) fatal("unable to create selfpipe");
  coe(selfpipe[0]);
  coe(selfpipe[1]);
  ndelay_on(selfpipe[0]);
  ndelay_on(selfpipe[1]);

  sig_block(sig_child);
  sig_catch(sig_child, s_child);
  sig_block(sig_term);
  sig_catch(sig_term, s_term);

  if (chdir(dir) == -1) fatal("unable to change to directory");
  svd[0].pid =0;
  svd[0].state =S_DOWN;
  svd[0].ctrl =C_NOOP;
  svd[0].want =W_UP;
  svd[0].islog =0;
  svd[1].pid =0;
  taia_now(&svd[0].start);
  if (stat("down", &s) != -1) svd[0].want =W_DOWN;

  if (stat("log", &s) == -1) {
    if (errno != error_noent)
      warn("unable to stat() ./log: ");
  }
  else {
    if (! S_ISDIR(s.st_mode))
      warnx("./log", 0, ": not a directory.");
    else {
      haslog =1;
      svd[1].state =S_DOWN;
      svd[1].ctrl =C_NOOP;
      svd[1].want =W_UP;
      svd[1].islog =1;
      taia_now(&svd[1].start);
      if (stat("log/down", &s) != -1)
        svd[1].want =W_DOWN;
      if (pipe(logpipe) == -1)
        fatal("unable to create log pipe");
      coe(logpipe[0]);
      coe(logpipe[1]);
    }
  }

  if (mkdir("supervise", 0700) == -1) {
    if ((r =readlink("supervise", buf, 256)) != -1) {
      if (r == 256)
        fatalx("unable to readlink ./supervise: ", "name too long");
      buf[r] =0;
      mkdir(buf, 0700);
    }
    else {
      if ((errno != ENOENT) && (errno != EINVAL))
        fatal("unable to readlink ./supervise");
    }
  }
  if ((svd[0].fdlock =open_append("supervise/lock")) == -1)
    fatal("unable to open supervise/lock");
  if (lock_exnb(svd[0].fdlock) == -1) fatal("unable to lock supervise/lock");
  coe(svd[0].fdlock);
  if (haslog) {
    if (mkdir("log/supervise", 0700) == -1) {
      if ((r =readlink("log/supervise", buf, 256)) != -1) {
        if (r == 256)
          fatalx("unable to readlink ./log/supervise: ", "name too long");
        buf[r] =0;
        if ((fd =open_read(".")) == -1)
          fatal("unable to open current directory");
        if (chdir("./log") == -1)
          fatal("unable to change directory to ./log");
        mkdir(buf, 0700);
        if (fchdir(fd) == -1)
          fatal("unable to change back to service directory");
        close(fd);
      }
      else {
        if ((errno != ENOENT) && (errno != EINVAL))
          fatal("unable to readlink ./log/supervise");
      }
    }
    if ((svd[1].fdlock =open_append("log/supervise/lock")) == -1)
      fatal("unable to open log/supervise/lock");
    if (lock_ex(svd[1].fdlock) == -1)
      fatal("unable to lock log/supervise/lock");
    coe(svd[1].fdlock);
  }

  fifo_make("supervise/control", 0600);
  if (stat("supervise/control", &s) == -1)
    fatal("unable to stat supervise/control");
  if (!S_ISFIFO(s.st_mode))
    fatalx("supervise/control exists but is not a fifo", "");
  if ((svd[0].fdcontrol =open_read("supervise/control")) == -1)
    fatal("unable to open supervise/control");
  coe(svd[0].fdcontrol);
  if ((svd[0].fdcontrolwrite =open_write("supervise/control")) == -1)
    fatal("unable to open supervise/control");
  coe(svd[0].fdcontrolwrite);
  update_status(&svd[0]);
  if (haslog) {
    fifo_make("log/supervise/control", 0600);
    if (stat("supervise/control", &s) == -1)
      fatal("unable to stat log/supervise/control");
    if (!S_ISFIFO(s.st_mode))
      fatalx("log/supervise/control exists but is not a fifo", "");
    if ((svd[1].fdcontrol =open_read("log/supervise/control")) == -1)
      fatal("unable to open log/supervise/control");
    coe(svd[1].fdcontrol);
    if ((svd[1].fdcontrolwrite =open_write("log/supervise/control")) == -1)
      fatal("unable to open log/supervise/control");
    coe(svd[1].fdcontrolwrite);
    update_status(&svd[1]);
  }
  fifo_make("supervise/ok",0600);
  if ((fd =open_read("supervise/ok")) == -1)
    fatal("unable to read supervise/ok");
  coe(fd);
  if (haslog) {
    fifo_make("log/supervise/ok",0600);
    if ((fd =open_read("log/supervise/ok")) == -1)
      fatal("unable to read log/supervise/ok");
    coe(fd);
  }
  for (;;) {
    iopause_fd x[3];
    struct taia deadline;
    struct taia now;
    char ch;

    if (haslog)
      if (! svd[1].pid && (svd[1].want == W_UP)) startservice(&svd[1]);
    if (! svd[0].pid)
      if ((svd[0].want == W_UP) || (svd[0].state == S_FINISH))
        startservice(&svd[0]);

    x[0].fd =selfpipe[0];
    x[0].events =IOPAUSE_READ;
    x[1].fd =svd[0].fdcontrol;
    x[1].events =IOPAUSE_READ;
    if (haslog) {
      x[2].fd =svd[1].fdcontrol;
      x[2].events =IOPAUSE_READ;
    }
    taia_now(&now);
    taia_uint(&deadline, 3600);
    taia_add(&deadline, &now, &deadline);

    sig_unblock(sig_term);
    sig_unblock(sig_child);
    iopause(x, 2 +haslog, &deadline, &now);
    sig_block(sig_term);
    sig_block(sig_child);

    while (read(selfpipe[0], &ch, 1) == 1)
      ;
    for (;;) {
      int child;
      int wstat;

      child =wait_nohang(&wstat);
      if (!child) break;
      if ((child == -1) && (errno != error_intr)) break;
      if (child == svd[0].pid) {
        svd[0].pid =0;
        pidchanged =1;
        svd[0].wstat =wstat;
        svd[0].ctrl &=~C_TERM;
        if (svd[0].state != S_FINISH)
          if ((fd =open_read("finish")) != -1) {
            close(fd);
            svd[0].state =S_FINISH;
            update_status(&svd[0]);
            continue;
          }
        svd[0].state =S_DOWN;
        taia_uint(&deadline, 1);
        taia_add(&deadline, &svd[0].start, &deadline);
        taia_now(&svd[0].start);
        update_status(&svd[0]);
        if (taia_less(&svd[0].start, &deadline)) sleep(1);
      }
      if (haslog) {
        if (child == svd[1].pid) {
          svd[1].pid =0;
          pidchanged =1;
          svd[1].state =S_DOWN;
          svd[1].ctrl &=~C_TERM;
          taia_uint(&deadline, 1);
          taia_add(&deadline, &svd[1].start, &deadline);
          taia_now(&svd[1].start);
          update_status(&svd[1]);
          if (taia_less(&svd[1].start, &deadline)) sleep(1);
        }
      }
    }
    if (read(svd[0].fdcontrol, &ch, 1) == 1) ctrl(&svd[0], ch);
    if (haslog)
      if (read(svd[1].fdcontrol, &ch, 1) == 1) ctrl(&svd[1], ch);

    if (sigterm) { ctrl(&svd[0], 'x'); sigterm =0; }

    if ((svd[0].want == W_EXIT) && (svd[0].state == S_DOWN)) {
      if (svd[1].pid == 0) _exit(0);
      if (svd[1].want != W_EXIT) {
        svd[1].want =W_EXIT;
        /* stopservice(&svd[1]); */
        update_status(&svd[1]);
        if (close(logpipe[1]) == -1) warn("unable to close logpipe[1]");
        if (close(logpipe[0]) == -1) warn("unable to close logpipe[0]");
      }
    }
  }
  _exit(0);
}
